src/osp/Active/SysNewton.cpp

```diff
@@ -486,18 +492,34 @@ void SysNewton::body_apply_torque(ACompRigidBody_t &body, Vector3 torque) noexce
     body.m_netTorque += torque;
 }
 
-Vector4 SysNewton::compute_body_CoM(ActiveScene& rScene,
-    ActiveEnt root, Matrix4 currentTransform)
+/* Since masses are usually stored in the rigidbody's children instead of the
+ * rigidbody root, the function provides the option to ignore the root entity's
+ * mass (this is the default). Otherwise, the root of each call will be double
+ * counted since it will also be included in both the loop and the root of the
+ * next call. This makes it possible for a rigidbody to have a convenient
+ * root-level mass component set by some external system without interfering
+ * with physics calculations that depend on mass components to determine the
+ * mechanical behaviors of the rigidbody.
+ */
+Vector4 SysNewton::compute_hier_CoM(ActiveScene& rScene,
+    ActiveEnt root, bool includeRootMass)
 {
     Vector3 localCoM{0.0f};
     float localMass = 0.0f;
 
-    for (ActiveEnt nextChild = root; nextChild != entt::null;)
+    if (includeRootMass)
     {
-        auto const &childHeir = rScene.reg_get<ACompHierarchy>(nextChild);
+        auto const* rootMass = rScene.reg_try_get<ACompMass>(root);
+        if (rootMass != nullptr) { localMass += rootMass->m_mass; }
+    }
+
+    for (ActiveEnt nextChild = rScene.reg_get<ACompHierarchy>(root).m_childFirst;
+        nextChild != entt::null;)
```

Comment:
wasn't there a range-based for-loop function for a hierachy that we could use?

Or is this recursing down a tree structure?

Ah, yea it is.

Can you check to see if (maybe even just ask directly) there's any utility functions in EnTT that'll make this sexier? like some kind of special notation that we can use to tell EnTT that we want a Tree structure, and then a way to ask for pre-order /  post order / level-order / so on.

Reply:
EnTT has no concept of trees, that's up to us. Which part did you want to change specifically, the `if` condition?

Reply:
Hrmm, that's really frustrating. It *should* have the concept of a hierarchy.

My goal was to say that the notion of traversing a tree, in the way that you're doing here, should really be something that the library provides us, instead of something we have to do manually.

For example, you're doing
```
        nextChild = childHier.m_siblingNext;
```

at the end of the loop, but it would be substantially easier for a reader of the code to understand what's going on if the concept of traversing a tree, the way you're doing here, was encapsulated inside some kind of iterator pair (or an iterator, and a c++20 iterator-sentinal) that internally embodied the tree traversal order.

Reply:
To be honest, it really shouldn't since ECS is supposed to be a framework removed from any concepts like that. It's not EnTT's job to deal with scene graphs. A fancy iterator scheme might be doable though.

Reply:
EnTT has far far more features than simply being dumb storage. There's no reason why it shouldn't provide basic "relationship traversal" capabilities out of the box, specifically because it has direct access to it's own internal design details that make it easier to write properly optimized traversal tools.